### ISSUE TITLE
Debugging ATtiny1616

### ISSUE THREAD

**Issue by bradanlane** [technoblogy/tiny-tft-graphics-library#1]
tl:dr; Are there any debugging techniques you use when a display doesn't show any output?

I am attempting to run **tiny-ttf-graphics-display** on an ATTiny1616. This is similar to the ATtiny412 but with _more stuff_.
The display is the 320x172 "rounded corner" ST7789 from Adafruit.
My configuration is:
```
int const dc = 1;
int const mosi = 2;
int const sck = 0;
int const cs = 3;

#define PORT_TOGGLE(x)  PORTC.OUTTGL = (x)
#define PORT_LOW(x)     PORTC.OUTCLR = (x)
#define PORT_HIGH(x)    PORTC.OUTSET = (x)
#define PORT_OUTPUT(x)  PORTC.DIRSET = (x)
```
These are the alternate pins for the SPI. Therefore, I added one line to `InitDisplay()` ...
`PORTMUX.CTRLB = PORTMUX_SPI0_ALTERNATE_gc;`

When I scope the SCK and MISO, the time results looks as expected. The clock is at approximately 2.5MHz.

Any debugging suggestions?

Are there specific settings required (Arduino: Tools ...)
![screenshot_20230523_155615](https://github.com/technoblogy/tiny-tft-graphics-library/assets/14001599/52ac9300-3cfd-4347-8f23-352e154a5378)

**Comment by technoblogy** [comment]
What you're doing looks correct. Does anything at all happen to the display? For example, when you call `InitDisplay()` does the display show a random pattern of bits?

Have you tried connecting the Reset pin to an I/O pin, and connecting it to Gnd briefly before running the library? Some displays are fussy about reset.

I think I've got one of those displays so I could try it if you're still stuck.

**Comment by bradanlane** [comment]
I see absolutely nothing on the display (other than the backlight). 

You suggest connecting `reset` to GND. I will try that. Currently it is connected through a pull-up resistor to VCC as shown in your blog post.

**Comment by technoblogy** [comment]
> You suggest connecting reset to GND. 

I mean taking it briefly to ground. See:

https://learn.adafruit.com/adafruit-1-47-round-rectangle-tft-display?view=all

**Comment by bradanlane** [comment]
OK. I will connect `reset` to a pin and pull it to GND (and then release it) prior to calling InitDisplay().

**Comment by technoblogy** [comment]
Let me know how you get on!

**Comment by bradanlane** [comment]
- Tested the display using other an ESP32S2 and Arduino+Adafruit libraries and it is good.
- Tested using ATtiny1616 and megatinycore+Adafruit libraries and nothing.
- Tested using ATtiny1616 and megatinycore+Tiny TFT Graphics Library v5 and nothing.

I do not currently have an ATTiny1616 with the default SPI pins accessible for testing.

**Comment by bradanlane** [comment]
If you are up to taking a look, I'm happy to send you one of the boards I am using.

**Comment by technoblogy** [comment]
One thing I've noticed - you've used PC1 for the DC signal, but that's the MISO pin for the alternate SPI, so there may be a conflict. Have you got a different pin on PORTC available you can try for DC?

**Comment by bradanlane** [comment]
The attiny1616 only has PC0 thru PC3. The only remaining pin is PB5.

My last option is order the Adafruit attiny1616 seesaw board for prototyping.

Other than the attiny85 and attiny412, which other chips have you or others had success?

**Comment by bradanlane** [comment]
Avoiding PC1 Worked!

**Comment by technoblogy** [comment]
> The attiny1616 only has PC0 thru PC3. The only remaining pin is PB5.

True. Why are you having to use the alternate SPI pins? If you were using the default SPI pins you'd have other pins on PORTA available. What board are you using at the moment?

> My last option is order the Adafruit attiny1616 seesaw board for prototyping.

Why not get an ATtiny1616 in an SOIC package and put it on a 20-pin SOIC breakout board?

**Comment by bradanlane** [comment]
My test board is the **ACK1** which has 42 LEDs, a piezo buzzer, Serial UART I/O, and 2 buttons.  So, all of PIN_PAn are used, and most of PIN_PBn us used.

When I start designing the actual PCB design, I will use PIN_PAn.


_(RE: using a SOIC, when it's all said and done, the Aidafruit seesaw is easier and with about $2)_

**Comment by technoblogy** [comment]
> Avoiding PC1 Worked!

Great!

Ah, OK, I've just found the ACK1, and I see it doesn't give you access to the standard SPI pins.

Yes, an Adafruit Seesaw seems like a good board, although I assume you won't be using the "Seesaw" software.

**Comment by bradanlane** [comment]
Now that I have the code working, I don't need another prototyping board. I was using the ACK1 because its' on my desk. I'll design a new board, now that this is working.

Thank you for your patience as I worked though this.

**Comment by technoblogy** [comment]
You're welcome. Perhaps I should add a note that you should leave the MISO pin free.

**Comment by technoblogy** [comment]
I've just realised that this library you've been using doesn't actually use the microcontroller's SPI peripheral; it controls the display via direct I/O pin manipulations. So you can use any four pins you like, as long as they are all in the same port, and you don't need to do:

    PORTMUX.CTRLB = PORTMUX_SPI0_ALTERNATE_gc;

I have published another display library that _does_ use SPI, and I got them confused when I gave you advice about this one.

With the ACK1 board you should therefore be able to use PC0, PC1, PC2, and PC3, but don't do the PORTMUX call, and I think it should work.

**Comment by bradanlane** [comment]
I looked for a display library that used _real SPI_. I didn't find it.

**Comment by bradanlane** [comment]
> With the ACK1 board you should therefore be able to use PC0, PC1, PC2, and PC3, but don't do the PORTMUX call, and I think it should work.

With or without `PORTMUX`, using `PC1` as `CS`, does not work.

**Comment by technoblogy** [comment]
> I looked for a display library that used real SPI. I didn't find it.

[Compact TFT Graphics Library](http://www.technoblogy.com/show?2LMJ)

**Comment by bradanlane** [comment]
Thanks. I was thinking you meant you had an _direct_ SPI implementation. I will test `Compact TFT Graphics Library` but will likely stay with this one since it has no dependencies on any Arduino framework or code - aka it will work with nothing more than the Microchip header files.

**Comment by technoblogy** [comment]
> With or without PORTMUX, using PC1 as CS, does not work.

I assume you mean DC. I can understand why it wouldn't work on the Adafruit library, because that uses SPI, but my Tiny TFT Graphics Library doesn't, so I can't think why it shouldn't work.

**Comment by technoblogy** [comment]
> but will likely stay with this one since it has no dependencies on any Arduino framework or code

In addition, it's about twice as fast, so it's the one I'd recommend for supported AVR processors.

**Comment by bradanlane** [comment]
My working code uses ...

>CS = PORTB_PIN5
>DS = PORTC_PIN3
>MOSI = PORTC_PIN2
>SCLK = PORTC_PIN0

According to the datasheet, DS, MOSI, and SCLK use the above pins on PORTC with PORTMUX to ALTERNATE.

**Comment by technoblogy** [comment]
> CS = PORTB_PIN5

I don't understand how that's working, because my macro `PORT_TOGGLE(x)` assumes the pins are all in PORTC. Or have you edited those references in the program?

**Comment by bradanlane** [comment]
all of your macros assume everything was on the same PORT and that is not possible so CS uses a new macro that just uses it's own PORT.

Given the way the code worked, I should have been able to use all PORTC pins but clearly, something in the chip disagreed.

**Comment by technoblogy** [comment]
Ah, so you changed those macro calls - understood. Glad it's all working!

**Comment by bradanlane** [comment]
As a point of reference, the `Compact TFT Graphics Library` is actually running slower than `tiny-tft-graphics-library`.

The only changes I made were to change the pins, add `SPI.swap(1)` to indicate the alternate pins, and to remove `random()` and replace it with simple maths.

I ran them side-by-side:
- `compact` takes 0.5s longer to initialize
- `compact` takes 0.2s longer to render the BarChart.

**Comment by technoblogy** [comment]
That's what I'd expect; sorry, I meant that the Tiny TFT Graphics Library is twice as fast as the Compact TFT Graphics Library.
